### ISSUE TITLE
Restriction lint for unsealed traits

### ISSUE THREAD

**Issue by Manishearth** [rust-lang/rust-clippy#14006]
A pattern in Rust is ["sealing" traits](https://predr.ag/blog/definitive-guide-to-sealed-traits-in-rust/): making them unimplementable to other users. It's typically done by doing something like:

```rust
pub trait PublicTrait: private::Sealed {
  // items
}

mod private {
   pub trait Sealed {} // empty
}
```

This allows one to declare that a trait is only used with internal types and should not be implemented by external users.

This has some benefits for keeping semver compatability. Codebases (like [ICU4X](https://github.com/unicode-org/icu4x)) may wish to make this the _default_, just like you can already opt in to `clippy::exhaustive_structs` and `clippy::exhaustive_enums` to avoid committing to too much semver wise with public structs/enums.

The idea would be a lint `clippy::unsealed_trait`, which lints on traits which are:

 - Public
 - _Reachable_
 - Do not depend (transitively) on any trait named `Sealed` (we could also ignore the naming) that are not reachable from other crates 
 - (perhaps) do not have a blanket impl for `T`

Transitivity might be tricky to do in a performant way.
